core/src/storage/collections/btree_map/impls.rs

```diff
@@ -0,0 +1,1872 @@
+enum InternalEntry<K, V> {
+    /// A vacant entry pointing to the next vacant index.
+    Vacant(Option<u32>),
+    /// An occupied entry contains a tree node with its elements.
+    Occupied(Node<K, V>),
+}
+
+/// An entry of a storage map.
+///
+/// This can either store the entries key/value pair or represent an entry that was
+/// removed after it has been occupied with key and value.
+#[derive(Encode, Decode)]
+pub enum Entry<'a, K, V> {
+    /// A vacant entry pointing to the next vacant index.
+    Vacant(VacantEntry<'a, K, V>),
+    /// An occupied entry containing the value.
+    Occupied(OccupiedEntry<'a, K, V>),
+}
+
+/// An occupied entry of a storage map.
+pub struct OccupiedEntry<'a, K, V> {
+    tree: &'a mut BTreeMap<K, V>,
+    handle: KVHandle,
+}
+
+impl<'a, K, V> VacantEntry<'a, K, V>
+where
+    K: Encode + Decode + Ord,
+    V: Encode + Decode,
+{
+    /// Gets a reference to the key that would be used when inserting a value
+    /// through the VacantEntry.
+    ///
+    /// # Examples
+    ///
+    /// ```no_compile
+    /// use ink_core::storage::BTreeMap;
+    ///
+    /// let mut map: BTreeMap<&str, usize> = new_btree_map();
+    /// assert_eq!(map.entry("poneyland").key(), &"poneyland");
+    /// ```
+    pub fn key(&self) -> &K {
+        self.key
+            .as_ref()
+            .expect("entry does always have a key; qed")
+    }
+
+    /// Sets the value of the entry with the `VacantEntry`'s key,
+    /// and returns a mutable reference to it.
+    ///
+    /// # Examples
+    ///
+    /// ```no_compile
+    /// use ink_core::storage::BTreeMap;
+    ///
+    /// let mut count: BTreeMap<&str, usize> = new_btree_map();
+    ///
+    /// // count the number of occurrences of letters in the vec
+    /// for x in vec!["a","b","a","c","a","b"] {
+    ///     *count.entry(x).or_insert(0) += 1;
+    /// }
+    ///
+    /// assert_eq!(count["a"], 3);
+    /// ```
+    pub fn insert(mut self, val: V) -> &'a mut V {
+        let key = self
+            .key
+            .take()
+            .expect("key is only taken when inserting, so must be there; qed");
+        if self.tree.is_empty() && self.tree.root().is_none() {
+            self.tree.create_root(key, val)
+        } else {
+            self.tree.insert_kv(self.handle, key, val)
+        }
+    }
+}
+
+impl<'a, K, V> OccupiedEntry<'a, K, V>
+where
+    K: Encode + Decode + Ord,
+    V: Encode + Decode,
+{
+    /// Gets a reference to the key in the entry.
+    ///
+    /// # Examples
+    ///
+    /// ```no_compile
+    /// use ink_core::storage::BTreeMap;
+    ///
+    /// let mut map: BTreeMap<&str, usize> = new_btree_map();
+    /// map.entry("poneyland").or_insert(12);
+    /// assert_eq!(map.entry("poneyland").key(), &"poneyland");
+    /// ```
+    pub fn key(&self) -> &K {
+        self.tree
+            .get_kv(self.handle)
+            .expect("every occupied entry always has a key/value pair; qed")
+            .0
```

Comment:
I'd love to see named structs instead of tuples or tuple structs in such cases since something like `.1` is not very readable tbh.